[HIPIFY] Make the cmake build system more friendly

ChrisKitching opened the pull request:
A few different things, broadly getting us to the position we agreed we
"eventually" wanted over in #230. Oooor you could just ignore the LLVM
changes and take the other useful stuff that's here :wink:

See commit messages for details, as usual.

emankov commented:
Chris, why don't you include one or a couple commits in one pull request? It'll be pretty easier for me to review them?

emankov commented on hipify-clang/README.md:
```diff
@@ -9,67 +13,67 @@
+were compiling the input file. The [Clang manual for compiling CUDA](https://llvm.org/docs/CompileCudaWithLLVM.html#compiling-cuda-code)
+may be useful.
+
+`-x cuda` may be omitted if the input file has a `.cu` or `.cuh` file extension.
```
It's just turned out that `-x cuda` may not be omitted.

ChrisKitching commented:
> Chris, why don't you include one or a couple commits in one pull request? It'll be pretty easier for me to review them?

Having a large number of small commits is intended to aid review. After all, the line delta here is less than 100 - but it's in many commits to make it clear what's going on.

If you step through the "commits" tab rather than the "files changed" tab, it should be straightforward to review (or, at least, define the subset you want to keep).

I aim to make pull requests that map nicely to logical units of work, are mutually orthogonal, and composed of well-defined commits. Hopefully :D. Filing something like this as individual PRs would take many days, since they are interdependent.


> It's just turned out that -x cuda may not be omitted.

How handy.

emankov commented on hipify-clang/CMakeLists.txt:
```diff
@@ -1,31 +1,14 @@
-    message(STATUS "hipify-clang will not be built. To build it please specify absolute path to LLVM 3.8 or higher using HIPIFY_CLANG_LLVM_DIR")
-    return()
-endif()
+find_package(LLVM REQUIRED)
```
I believe we should leave for users an opportunity for specifying exact LLVM.

emankov replied:
I missed your words about CMAKE_PREFIX_PATH for exact LLVM. CMAKE_PREFIX_PATH is ok, but add these words to readme, please.

ChrisKitching replied:
> add these words to readme, please.

This was already done: https://github.com/ROCm-Developer-Tools/HIP/pull/236/commits/6a8dc2dd56949b386fb19bb27fc0b1d617047f24#diff-675eed3cdf4ce542842286f5722ee3bbR24

Perhaps this wording isn't clear enough? We can change it if this is confusing or something - I'm not exactly a great writer ;)

emankov replied:
So do I, let's leave it as it is written.

emankov commented on hipify-clang/CMakeLists.txt:
```diff
@@ -1,31 +1,14 @@
-    return()
-endif()
+find_package(LLVM REQUIRED)
+message(STATUS "Found LLVM ${LLVM_PACKAGE_VERSION}")
```
Let's say the place of the found LLVM.
`message(STATUS "Found LLVM ${LLVM_PACKAGE_VERSION} in ${LLVM_CMAKE_DIR}")`

emankov commented on CMakeLists.txt:
```diff
@@ -2,6 +2,11 @@ cmake_minimum_required(VERSION 2.8.3)
 #############################
+# Options
+#############################
+option(BUILD_HIPIFY_CLANG "Enable building the CUDA->HIP converter" ON)
```
Make BUILD_HIPIFY_CLANG OFF by default as it was before, moreover this change breaks HIP build on Jenkins due to LLVM not found.

```
[hcc_ctu] CMake Error at hipify-clang/CMakeLists.txt:4 (find_package):
[hcc_ctu]   By not providing "FindLLVM.cmake" in CMAKE_MODULE_PATH this project has
[hcc_ctu]   asked CMake to find a package configuration file provided by "LLVM", but
[hcc_ctu]   CMake did not find one.
[hcc_ctu] 
[hcc_ctu]   Could not find a package configuration file provided by "LLVM" with any of
[hcc_ctu]   the following names:
[hcc_ctu] 
[hcc_ctu]     LLVMConfig.cmake
[hcc_ctu]     llvm-config.cmake
[hcc_ctu] 
[hcc_ctu]   Add the installation prefix of "LLVM" to CMAKE_PREFIX_PATH or set
[hcc_ctu]   "LLVM_DIR" to a directory containing one of the above files.  If "LLVM"
[hcc_ctu]   provides a separate development package or SDK, be sure it has been
[hcc_ctu]   installed.
[hcc_ctu] 
[hcc_ctu] 
[hcc_ctu] -- Configuring incomplete, errors occurred!
```

ChrisKitching commented:
All suggested changes applied, thanks for spotting those issues.

- hipify is default-off now
- The remark about `-x cuda` has been removed

I went slightly further than just printing the LLVM cmake module path, and had it print the include and binary paths, too. Just to make it completely clear exactly what we're using:

```
-- Found LLVM 5.0.0:
--    - CMake module path: /usr/lib/cmake/llvm
--    - Include path     : /usr/include
--    - Binary path      : /usr/bin
```

Let me know if I missed a spot!

emankov commented on CMakeLists.txt:
```diff
@@ -2,6 +2,11 @@ cmake_minimum_required(VERSION 2.8.3)
 #############################
+# Options
+#############################
+option(BUILD_HIPIFY_CLANG "Enable building the CUDA->HIP converter" OFF
```
typo: missing `)`, HIP build is still broken.

ChrisKitching commented:
> LGTM, just fixed a typo.

Wow, that was an epic fail on my part ;). Rebased it into the feature branch now. No  need to pollute the git history with a non-bisectable range because I can't use my keyboard!

emankov commented:
Sorry, Chris, but I've just wanted to finally merge this PR at the end of the week. Editing the file was triggered Jenkins testing, which logs I can observe only at work place. Well, let's wait Jenkins awakening once again :)

ChrisKitching commented:
> Sorry, Chris, but I've just wanted to finally merge this PR at the end of the week.

Don't worry about it! Definitely better to be slow and correct!

I realise I have a tendency to overwhelm reviewers, so I sort of go in expecting things to sit around for a while. It's no big deal - I don't want to break stuff :grin:

> Editing the file has triggered Jenkins testing, which logs I can observe only at work place. Well, let's wait Jenkins awakening once again :)

Sounds like you guys need to get yourselves a VPN!

emankov commented:
Chris, I'm going on vacation next week till 15th of November, so please add also Ben and Daniil as reviewers of your upcomming changes in hipify, and Ben and Maneesh if HIP is touched. I'll be looking at your very important contribution in hipify also and may be will say smth :) and even merge, but with delays.

ChrisKitching commented:
> Chris, I'm going on vacation next week till 15th of November

Have fun!
